src/main/kotlin/sc/iview/ui/SwingMainWindow.kt

```diff
@@ -58,228 +59,239 @@ class SwingMainWindow(val sciview: SciView) : MainWindow {
-            nodePropertyEditor.rebuildTree()
-            logger.info("Done initializing SciView")
+            })
+
+            frame.isVisible = true
+            glassPane.repaint()
+
+            sciview.sceneryPanel[0] = sceneryJPanel
+            val renderer = Renderer.createRenderer(
+                    sciview.hub,
+                    sciview.applicationName,
+                    sciview.currentScene,
+                    sciview.windowWidth,
+                    sciview.windowHeight,
+                    sciview.sceneryPanel[0]
+            )
+
+            sciview.setRenderer(renderer)
+            sciview.hub.add(SceneryElement.Renderer, renderer)
+            sciview.reset()
+
+            SwingUtilities.invokeLater {
+                try {
+                    while (!sciview.getSceneryRenderer()!!.firstImageReady) {
+                        logger.debug("Waiting for renderer initialisation")
+                        Thread.sleep(300)
+                    }
+                    Thread.sleep(200)
+                } catch (e: InterruptedException) {
+                    logger.error("Renderer construction interrupted.")
+                }
+                nodePropertyEditor.rebuildTree()
+                logger.info("Done initializing SciView")
 
-            // subscribe to Node{Added, Removed, Changed} events, happens automatically
+                // subscribe to Node{Added, Removed, Changed} events, happens automatically
 //            eventService!!.subscribe(this)
-            sceneryJPanel.isVisible = true
+                sceneryJPanel.isVisible = true
 
-            // install hook to keep inspector updated on external changes (scripting, etc)
-            sciview.currentScene.onNodePropertiesChanged["updateInspector"] = { node: Node ->
-                if (node === nodePropertyEditor.currentNode) {
-                    nodePropertyEditor.updateProperties(node)
+                // install hook to keep inspector updated on external changes (scripting, etc)
+                sciview.currentScene.onNodePropertiesChanged["updateInspector"] = { node: Node ->
+                    if (node === nodePropertyEditor.currentNode) {
+                        nodePropertyEditor.updateProperties(node)
+                    }
                 }
-            }
 
-            // Enable push rendering by default
-            renderer.pushMode = true
-            sciview.camera!!.setPosition(1.65, 1)
-            glassPane.isVisible = false
+                // Enable push rendering by default
+                renderer.pushMode = true
+                sciview.camera!!.setPosition(1.65, 1)
+                glassPane.isVisible = false
+
+                sceneryJPanel.minimumSize = Dimension(256, 256)
+            }
+        }
 
-            sceneryJPanel.minimumSize = Dimension(256, 256)
+        if(SwingUtilities.isEventDispatchThread()) {
+            initializer.run()
+        } else {
+            SwingUtilities.invokeAndWait(initializer)
```

Comment:
The SciJava `ThreadService` [already has such case logic](https://github.com/scijava/scijava-common/blob/scijava-common-2.94.1/src/main/java/org/scijava/thread/DefaultThreadService.java#L108-L120), so you can do:
```kotlin
sciview.scijavaContext.service(ThreadService::class.java).invoke(initializer)
```

Reply:
Thanks, I'll also address this post-PR